src/Form/UserPickupRequestForm.php

```diff
@@ -0,0 +1,413 @@
+<?php
+namespace Drupal\arborcat\Form;
+
+use Drupal\Core\Form\FormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Mail\MailManagerInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Predis\Client;
+use Drupal\arborcat\Controller;
+use DateTime;
+use DateTimeHelper;
+
+class UserPickupRequestForm extends FormBase
+{
+    public function getFormId()
+    {
+        return 'user_pickup_request_form';
+    }
+
+    public function buildForm(array $form, FormStateInterface $form_state, string $patronId = null, string $requestLocation = null, string $mode = null)
+    {
+        $guzzle = \Drupal::httpClient();
+        $api_key = \Drupal::config('arborcat.settings')->get('api_key');
+        $api_url = \Drupal::config('arborcat.settings')->get('api_url');
+
+        $patron_info = json_decode((string) $guzzle->get("$api_url/patron?apikey=$api_key&pnum=$patronId")->getBody()->getContents(), true);
+
+        $uid = $patron_info['evg_user']['card']['id'];
+        $account = \Drupal\user\Entity\User::load($uid);
+
+        $api_key = $account->get('field_api_key')->value;
+
+        $patron_barcode = $patron_info['evg_user']['card']['barcode'];
+
+        $selfCheckApi_key = \Drupal::config('arborcat.settings')->get('selfcheck_key');
+        $selfCheckApi_key .= '-' .  $patron_barcode;
+        $patron_holds = json_decode($guzzle->get("$api_url/patron/$selfCheckApi_key/holds")->getBody()->getContents(), true);
+
+        $eligible_holds = [];
+
+        // Get the locations
+        $locations = json_decode($guzzle->get("$api_url/locations")->getBody()->getContents());
+        $locationName = $locations->$requestLocation;
+
+        // check the mode to see whether we need to display "Cancel Mode, rather than pickup request mode
+        if ('cancel' == $mode) {
+            // do cancel things
+        }
+        //start at 1 to avoid issue with eligible holds array not being zero-based
+        $i=1;
+
+        $mel_mappings = [
```

Comment:
Okay to hard code values to get things running, but could be candidate for a config lookup table, either in drupal settings or something on the API side